posts/inside-rust/2021-03-15-What-the-error-handling-project-group-is-working-towards.md

```diff
@@ -0,0 +1,201 @@
+---
+
+This blog post is a follow up of our [previous](https://blog.rust-lang.org/inside-rust/2020/11/23/What-the-error-handling-project-group-is-working-on.html) post detailing what we're working on now. We figure it's time to take a step back and answer some of the bigger questions of why we're making these changes and how they're going to fit into the bigger picture of error-handling provided by Rust.
+
+> Disclaimer: *This post is equal parts plan and aspiration. There are technical challenges here to sort out so the final outcome may look rather different from our initial vision, so please don't assume any of this is final.*
+
+## Guidelines for implementing `Display::fmt` and `Error::source`
+
+But before we get into the plans we first need to share an announcement. Project Error Handling recently created a guideline for [how to implement `Display::fmt` and `Error::source`](https://github.com/rust-lang/project-error-handling/issues/27#issuecomment-763950178). To create this guideline we first had to answer a fundamental question, _What is the primary role of the error trait?_.
+
+As we saw it there are two possible answers.
+
+* The error trait is primarily an interface for reporting errors.
+* The error trait is primarily an interface for reacting to specific errors.
+
+In the first case, "an interface for reporting errors" means it provides access to pieces of information about the error such as the error message, optional backtrace, optional source, and so on. A reporter then takes all this information and stitches it together into a full error report, explaining what went wrong.
+
+If we decide to prefer the former viewpoint then we would want to prioritize not duplicating information in the reporting interfaces of the `Error` trait. That would then turn into the guidance: _Return source errors via `Error::source` unless the source error's message is included in your own error message in `Display`._
+
+In the second case, "an interface for reacting to specific errors" is referring to the various `downcast` methods associated with the error trait. These methods can be used to test if a type erased error is a specific error variant and they're functionally equivalent to `match`.
+
+If we decide to prefer the latter then we would want to prioritize exposing every error type in the chain of errors to ensure users can always correctly downcast to specific error variants they want to handle. In this situation the guidance would be: _Always return source errors via `Error::source`._
+
+After some discussion we concluded that the reporting focused viewpoint was the ideal one. We prefer splitting our errors into multiple error messages with no duplication of information as opposed to having each error print itself followed by all others. This split moves the responsibility of formatting to the end user, rather than relying on many library authors concatenating error messages together with a consistent format.
```

Comment:
Why is this is considered the "ideal" approach? It feels as if the mission of the "The Error Handling Project Group" was to prioritise the presentation of errors rather than the handling of errors (as indicated in the project group name). For a systems programming language I feel that the handling of errors should have priority over the presentation of errors; there're systems which do not even have the ability to present error messages to a user (e.g. embedded devices) but where error handling is of utmost importance (and rather poorly supported by Rust at the moment).

Reply:
> Why is this considered the "ideal" approach?

Perhaps ideal isn't the best vocab here, since there isn't really a right answer, just a best one. The reasoning is that we evaluated the trade-offs and costs of the various options and found that this was the least bad option, given that it doesn't prevent reacting to source errors via downcasting, where-as the alternative completely prevents configurable and consistent error reporting formats by duplicating information throughout the `Error` interface. If you're interested in a more in-depth analysis of the trade offs and why we felt it was best to prioritize deduplicating information please checkout this issue: https://github.com/rust-lang/project-error-handling/issues/27

> It feels as if the mission of the "The Error Handling Project Group" was to prioritise the presentation of errors rather than the handling of errors (as indicated in the project group name). For a systems programming language I feel that the handling of errors should have priority over the presentation of errors; there're systems which do not even have the ability to present error messages to a user (e.g. embedded devices) but where error handling is of utmost importance (and rather poorly supported by Rust at the moment).

"Error Handling" a superset of responsibilities, which in my mind includes creating, propagating, matching/reacting, discarding, and reporting. This is more accurately a tension between how you can programmatically react to specific errors and how you represent and report them, as opposed to handling in general. I think its well within the mission of the error handling project group to analyze the trade offs and make a recommendation one direction or the other.